### PR TITLE
Backend/endpoints

### PR DESCRIPTION
**Pull request by pmh192**
- GET /api/tournament/getAllCreated
- GET /api/user/get
- GET /api/team/getAll
- POST /api/tournament/{tournamentID}/start
- GET /api/enum/tournament/status
- GET /api/tournament/{tournamentID}/match/getAll

### CODE REVIEW

File: backend/src/main/java/com/tourneynizer/tourneynizer/dao/MatchDao.java [hunk 1]
```diff
@@ -34,14 +40,14 @@ public void insert(Match match, User user) throws EmailTakenException, SQLExcept
...
                 PreparedStatement preparedStatement = connection.prepareStatement(sql, new String[]{"id"});
                 preparedStatement.setLong(1, match.getTournamentId());
-                preparedStatement.setLong(2, match.getTeam1_id());
-                preparedStatement.setLong(3, match.getTeam2_id());
-                preparedStatement.setLong(4, match.getScore1());
-                preparedStatement.setLong(5, match.getScore2());
+                setNullable(preparedStatement, 2, match.getTeam1Id());
+                setNullable(preparedStatement, 3, match.getTeam2Id());
+                setNullable(preparedStatement, 4, match.getScore1());
+                setNullable(preparedStatement, 5, match.getScore2());
```
Comment (lar-ryan-nick) [inline backend/src/main/java/com/tourneynizer/tourneynizer/dao/MatchDao.java]: A few things here:
* Why don't you store child 1 and child 2 in the database? Do you have some way of keeping track of it through match order?
* Please don't forget to remove court number at some point please.

File: backend/src/main/java/com/tourneynizer/tourneynizer/model/MatchGenerator.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        List<Match> matches = new ArrayList<>();
+
+        while (tree.size() >= 2) {
+            MatchNode node1 = tree.poll();
+            MatchNode node2 = tree.poll();
+
+            Match parent = new Match(tournament.getId(), order++, node1.getValueId(), node2.getValueId(), null, ScoreType.ONE_SET);
+            matchDao.insert(parent, user);
+            tree.add(new MatchNode(parent, orderInserted++, Math.max(node1.height, node2.height) + 1));
+            matches.add(parent);
```
Comment (lar-ryan-nick) [inline backend/src/main/java/com/tourneynizer/tourneynizer/model/MatchGenerator.java]: Ok this looks pretty cool, but are you also prepared for a consolation bracket? Or is it just easier to do a one and done?

File: backend/src/main/java/com/tourneynizer/tourneynizer/model/TournamentStatus.java [hunk 3]
```diff
@@ -0,0 +1,5 @@
+package com.tourneynizer.tourneynizer.model;
+
+public enum TournamentStatus {
+    CREATED, STARTED, FINISHED;
+}
```
Comment (lar-ryan-nick) [inline backend/src/main/java/com/tourneynizer/tourneynizer/model/TournamentStatus.java]: Good idea, but we could also keep track of this through the database values (I would add an actualStartTime and change the old startTime to plannedStartTime):
* Created - Actual startTime is null, but planned startTime has a value
* Started - Actual startTime has a value, but endTime is null
* Finished - endTime has a value

File: backend/src/main/java/com/tourneynizer/tourneynizer/controller/TournamentController.java [hunk 4]
```diff
@@ -65,4 +67,45 @@ public TournamentController(TournamentService tournamentService, SessionService
...
+            return new ResponseEntity<Object>(new ErrorMessage(e), new HttpHeaders(), HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    private static TournamentStatus[] values;
+    private synchronized static TournamentStatus[] getValues() {
+        if (values== null) {
+            values = TournamentStatus.values();
+        }
+        return values;
```
[the remark this answers is not in the document]
Comment (pmh192) [reply backend/src/main/java/com/tourneynizer/tourneynizer/controller/TournamentController.java]: Yeah in order to guarantee immutability, a new array is returned every time `values` is called.

File: backend/src/main/java/com/tourneynizer/tourneynizer/model/MatchGenerator.java (hunk 2 above)
Comment (pmh192) [reply backend/src/main/java/com/tourneynizer/tourneynizer/model/MatchGenerator.java]: This is easier; it's just a starting point.

File: backend/src/main/java/com/tourneynizer/tourneynizer/model/TournamentStatus.java (hunk 3 above)
Comment (pmh192) [reply backend/src/main/java/com/tourneynizer/tourneynizer/model/TournamentStatus.java]: I think having two different start times is a little too complex for right now.

File: backend/src/main/java/com/tourneynizer/tourneynizer/dao/MatchDao.java (hunk 1 above)
Comment (pmh192) [reply backend/src/main/java/com/tourneynizer/tourneynizer/dao/MatchDao.java]: - Forgot, will fix
- Soon

File: backend/src/main/java/com/tourneynizer/tourneynizer/model/Match.java [hunk 5]
```diff
@@ -49,28 +40,11 @@ public void setTournamentId(long tournament_id) {
...
-            throw new IllegalArgumentException("Teams cannot ref their own game");
-        }
-        this.team2Id = team2Id;
-    }
 
     public void setRefId(Long refId) {
-        if (refId != null && (Objects.equals(refId, team1Id) || Objects.equals(refId, team2Id))) {
+        long firstChild = matchChildren.first();
+        long secondChild = matchChildren.second();
+        if (refId != null && (Objects.equals(refId, firstChild) || Objects.equals(refId, secondChild))) {
```
Comment (ankushrayabhari) [inline backend/src/main/java/com/tourneynizer/tourneynizer/model/Match.java]: Here refId is a team ID and firstChild/secondChild are either match IDs or team IDs depending on their position. Update the first and second methods to make sure they return only the winner of the first and second match, not the match itself.

**Pull request merged**